File: renu_customization/renu_customization/page/margin_build/margin_build.py
import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate, date_diff, now_datetime
import json
import openpyxl
from openpyxl.styles import Alignment, Font, PatternFill, Border, Side
from openpyxl.utils import get_column_letter
from io import BytesIO
import base64
from renu_customization.renu_customization.report.sales_invoice_report.sales_invoice_report import execute

@frappe.whitelist()
def export_to_pdf(html):
    pdf_content = frappe.utils.pdf.get_pdf(html, {"orientation": "Landscape"})
    frappe.response.filename = f"Margin_Build_{nowdate()}.pdf"
    frappe.response.filecontent = pdf_content
    frappe.response.type = "download"

@frappe.whitelist()
def get_dashboard_data(filters=None):
    if isinstance(filters, str):
        filters = json.loads(filters)
    if not filters:
        filters = {}

    # Handle Fiscal Year
    if filters.get("fiscal_year"):
        fy = frappe.get_doc("Fiscal Year", filters.get("fiscal_year"))
        if fy:
            filters["from_date"] = fy.year_start_date
            filters["to_date"] = fy.year_end_date

    base_filters = frappe._dict({})
    if filters.get("from_date"):
        base_filters.from_date = filters.get("from_date")
    if filters.get("to_date"):
        base_filters.to_date = filters.get("to_date")

    report_result = execute(base_filters)
    raw_data = report_result[1] if report_result and len(report_result) > 1 else []

    data = []
    
    inv_names = list(set([d.get("invoice_id") or d.get("name") or d.get("parent") for d in raw_data if d.get("invoice_id") or d.get("name") or d.get("parent")])) if raw_data else []
    dom_exp_map = {}
    status_map = {}
    
    if inv_names:
        invoices = frappe.get_all("Sales Invoice", filters={"name": ("in", inv_names)}, fields=["name", "is_domestic", "is_export", "invoice_type", "status"])
        for i in invoices:
            status_map[i.name] = i.status
            if i.is_domestic: 
                dom_exp_map[i.name] = "Domestic"
            elif i.is_export: 
                dom_exp_map[i.name] = "Export"
            elif i.invoice_type and "Domestic" in i.invoice_type:
                dom_exp_map[i.name] = "Domestic"
            elif i.invoice_type and "Export" in i.invoice_type:
                dom_exp_map[i.name] = "Export"
            else: 
                dom_exp_map[i.name] = "Other"

    for row in raw_data:
        inv_id = row.get("invoice_id") or row.get("name") or row.get("parent")
        
        status = status_map.get(inv_id)
        if status == "Cancelled":
            continue

        revenue = flt(row.get("base_amount"))
        qty = flt(row.get("qty"))
        purchase_rate = flt(row.get("item_purchase_rate"))
        cogs = qty * purchase_rate
        margin = revenue - cogs
        
        row_type = dom_exp_map.get(inv_id, "Other")

        if filters.get("type") and row_type != filters.get("type"):
            continue

        if filters.get("customer") and filters.get("customer") != row.get("customer"):
            continue
            
        data.append({
            "name": inv_id,
            "posting_date": row.get("invoice_date") or row.get("posting_date"),
            "customer": row.get("customer"),
            "customer_name": row.get("customer_name"),
            "item_code": row.get("item_code"),
            "item_name": row.get("item_name"),
            "item": f"{row.get('item_code')} {row.get('item_name')}" if row.get('item_code') else (row.get('item_name') or ""),
            "type": row_type,
            "margin": margin,
            "revenue": revenue,
            "cogs": cogs
        })

    # No longer aggregating by invoice to show items
    final_results = data
    final_results.sort(key=lambda x: getdate(x["posting_date"]) if x["posting_date"] else getdate(), reverse=True)

    total_margin = 0
    export_margin = 0
    domestic_margin = 0
    
    unique_invoices = set()

    for d in final_results:
        amt = flt(d["margin"])
        total_margin += amt
        if d["type"] == "Export":
            export_margin += amt
        elif d["type"] == "Domestic":
            domestic_margin += amt
        
        unique_invoices.add(d["name"])

    summary = [
        {"label": _("Total Margin"), "value": total_margin, "indicator": "green", "fieldtype": "Currency"},
        {"label": _("Export Margin"), "value": export_margin, "indicator": "orange", "fieldtype": "Currency"},
        {"label": _("Domestic Margin"), "value": domestic_margin, "indicator": "blue", "fieldtype": "Currency"},
        {"label": _("Total Invoices"), "value": len(unique_invoices), "indicator": "purple", "fieldtype": "Int"}
    ]

    charts = {
        "margin_breakdown": {
            "title": _("Margin Breakdown (Export vs Domestic)"),
            "data": {
                "labels": [_("Export Margin"), _("Domestic Margin")],
                "datasets": [{"name": _("Margin"), "values": [export_margin, domestic_margin]}]
            },
            "type": "donut",
            "colors": ["#f59e0b", "#3b82f6"]
        }
    }

    return {
        "summary": summary,
        "charts": charts,
        "results": final_results
    }

@frappe.whitelist()
def export_to_excel(filters=None, export_type="all"):
    dashboard_data = get_dashboard_data(filters)
    data = dashboard_data.get("results")
    summary = dashboard_data.get("summary")
    
    if not data:
        return None

    wb = openpyxl.Workbook()
    
    # Based on export_type, we decide which sheets to keep
    if export_type == "all":
        ws_overview = wb.active
        ws_overview.title = "Dashboard Overview"
        ws_list = wb.create_sheet("Margin List")
    elif export_type == "detail":
        ws_list = wb.active
        ws_list.title = "Margin List"
        ws_overview = wb.create_sheet("Dummy") # Will be removed
    
    header_fill = PatternFill(start_color="2c3e50", fill_type="solid")
    header_font = Font(bold=True, color="FFFFFF")
    title_font = Font(bold=True, size=14)
    section_font = Font(bold=True, size=12)
    thin_side = Side(style='thin')
    table_border = Border(left=thin_side, right=thin_side, top=thin_side, bottom=thin_side)
    
    if export_type == "all":
        ws_overview.cell(row=1, column=1, value="Margin Build Dashboard (Million INR)").font = title_font
        ws_overview.cell(row=1, column=4, value="Generated On: " + now_datetime().strftime("%Y-%m-%d %H:%M"))
        
        ws_overview.cell(row=3, column=1, value="Margin Metrics Summary").font = section_font
        
        colors = {"red": "ef4444", "orange": "f59e0b", "blue": "3b82f6", "green": "10b981", "purple": "8b5cf6"}
        
        for i, s in enumerate(summary):
            r = 5 + (i // 2) * 3
            c = 1 + (i % 2) * 2
            bg_color = colors.get(s.get('indicator', 'blue').lower(), "3b82f6")
            
            cell_l = ws_overview.cell(row=r, column=c, value=s.get('label'))
            cell_l.font = Font(bold=True, color="FFFFFF")
            cell_l.fill = PatternFill(start_color=bg_color, fill_type="solid")
            cell_l.alignment = Alignment(horizontal="center")
            ws_overview.merge_cells(start_row=r, start_column=c, end_row=r, end_column=c+1)
            
            val = s.get('value')
            if s.get('fieldtype') == 'Currency':
                cell_v = ws_overview.cell(row=r+1, column=c, value=flt(val) / 1000000)
                cell_v.number_format = '"₹ "#,##0.0000" M"'
            else:
                cell_v = ws_overview.cell(row=r+1, column=c, value=val)
            cell_v.font = Font(bold=True, size=11)
            cell_v.alignment = Alignment(horizontal="center")
            cell_v.border = Border(bottom=Side(style='medium', color=bg_color))
            ws_overview.merge_cells(start_row=r+1, start_column=c, end_row=r+1, end_column=c+1)

    if export_type in ["all", "detail"]:
        row_idx = 1
        ws_list.cell(row=row_idx, column=1, value="Detailed Margin List").font = section_font
        row_idx += 2
        
        headers = ["S.No.", "Invoice ID", "Date", "Customer", "Item", "Type", "Revenue (M)", "COGS (M)", "Margin (M)"]
        for idx, h in enumerate(headers, start=1):
            cell = ws_list.cell(row=row_idx, column=idx, value=h)
            cell.font, cell.fill, cell.alignment, cell.border = header_font, header_fill, Alignment(horizontal="center"), table_border
        row_idx += 1
        
        for r_idx, row in enumerate(data):
            ws_list.cell(row=row_idx, column=1, value=r_idx + 1).border = table_border
            ws_list.cell(row=row_idx, column=2, value=row['name']).border = table_border
            ws_list.cell(row=row_idx, column=3, value=row['posting_date']).border = table_border
            ws_list.cell(row=row_idx, column=4, value=row['customer_name'] or row['customer']).border = table_border
            ws_list.cell(row=row_idx, column=5, value=row['item']).border = table_border
            ws_list.cell(row=row_idx, column=6, value=row['type']).border = table_border
            
            rev_cell = ws_list.cell(row=row_idx, column=7, value=flt(row['revenue']) / 1000000)
            rev_cell.number_format, rev_cell.border = '"₹ "#,##0.0000" M"', table_border
            
            cogs_cell = ws_list.cell(row=row_idx, column=8, value=flt(row['cogs']) / 1000000)
            cogs_cell.number_format, cogs_cell.border = '"₹ "#,##0.0000" M"', table_border
            
            margin_cell = ws_list.cell(row=row_idx, column=9, value=flt(row['margin']) / 1000000)
            margin_cell.number_format, margin_cell.border = '"₹ "#,##0.0000" M"', table_border
            
            row_idx += 1
    
        # Add Grand Total Row for Margin List
        ws_list.cell(row=row_idx, column=1, value="Grand Total").font = header_font
        ws_list.merge_cells(start_row=row_idx, start_column=1, end_row=row_idx, end_column=6)
        for c in range(1, 7):
            ws_list.cell(row=row_idx, column=c).fill = header_fill
            ws_list.cell(row=row_idx, column=c).border = table_border
            
        total_rev = sum(flt(row['revenue']) for row in data) / 1000000
        c_rev = ws_list.cell(row=row_idx, column=7, value=total_rev)
        c_rev.number_format, c_rev.font, c_rev.fill, c_rev.border = '"₹ "#,##0.0000" M"', header_font, header_fill, table_border
        c_rev.alignment = Alignment(horizontal="right")
        
        total_cogs = sum(flt(row['cogs']) for row in data) / 1000000
        c_cogs = ws_list.cell(row=row_idx, column=8, value=total_cogs)
        c_cogs.number_format, c_cogs.font, c_cogs.fill, c_cogs.border = '"₹ "#,##0.0000" M"', header_font, header_fill, table_border
        c_cogs.alignment = Alignment(horizontal="right")
        
        total_margin = sum(flt(row['margin']) for row in data) / 1000000
        c_margin = ws_list.cell(row=row_idx, column=9, value=total_margin)
        c_margin.number_format, c_margin.font, c_margin.fill, c_margin.border = '"₹ "#,##0.0000" M"', header_font, header_fill, table_border
        c_margin.alignment = Alignment(horizontal="right")

    # Remove dummy if detail only
    if export_type == "detail" and "Dummy" in wb.sheetnames:
        wb.remove(wb["Dummy"])

    # Column Widths
    for ws in wb.worksheets:
        for i in range(1, 10):
            ws.column_dimensions[get_column_letter(i)].width = 25

    output = BytesIO()
    wb.save(output)
    output.seek(0)
    
    filename = f"Margin_Build_{nowdate()}.xlsx"
    if export_type == "detail":
        filename = f"Detailed_Margin_List_{nowdate()}.xlsx"
        
    return {
        "filename": filename,
        "filecontent": base64.b64encode(output.read()).decode()
    }

File: renu_customization/renu_customization/page/margin_build/margin_build.js
frappe.pages["margin_build"].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Margin Build (Million INR)"),
		single_column: true
	});

	page.set_primary_action(__("Refresh"), () => page.refresh());

	// Export Functions
	const export_to_excel = () => {
		const filters = page.filter_group.get_values();
        frappe.show_alert({ message: __("Generating Excel Report..."), indicator: "blue" });

		frappe.call({
			method: "renu_customization.renu_customization.page.margin_build.margin_build.export_to_excel",
			args: { filters: filters },
			callback: function (r) {
				if (r.message) {
					const { filename, filecontent } = r.message;
					const byteCharacters = atob(filecontent);
					const byteNumbers = new Array(byteCharacters.length);
					for (let i = 0; i < byteCharacters.length; i++) {
						byteNumbers[i] = byteCharacters.charCodeAt(i);
					}
					const byteArray = new Uint8Array(byteNumbers);
					const blob = new Blob([byteArray], {
						type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
					});
					const link = document.createElement("a");
					link.href = window.URL.createObjectURL(blob);
					link.download = filename;
					link.click();
                    frappe.show_alert({ message: __("Excel Report Downloaded"), indicator: "green" });
				}
			},
		});
	};

	const export_pdf = async () => {
		if (!page.dashboard_data) return;

		const report_date = moment().format("YYYY-MM-DD HH:mm");
		const data = page.dashboard_data;
		
		const get_chart_png = () => {
			const svg = document.querySelector(`#margin-chart svg`);
			if (!svg) return null;
			const canvas = document.createElement("canvas");
			const context = canvas.getContext("2d");
			const svg_data = new XMLSerializer().serializeToString(svg);
			const img = new Image();
			return new Promise((resolve) => {
				img.onload = () => {
					canvas.width = img.width * 2;
					canvas.height = img.height * 2;
					context.fillStyle = "white";
					context.fillRect(0, 0, canvas.width, canvas.height);
					context.drawImage(img, 0, 0, canvas.width, canvas.height);
					resolve(canvas.toDataURL("image/png"));
				};
				img.src = "data:image/svg+xml;base64," + btoa(unescape(encodeURIComponent(svg_data)));
			});
		};

		const chart_png = await get_chart_png();
        const total_val = data.charts.margin_breakdown.data.datasets[0].values.reduce((a, b) => a + b, 0) || 1;

        // Generate Legend Box HTML for PDF
        const legend_box_html = data.charts.margin_breakdown.data.labels.map((label, i) => {
            const val = data.charts.margin_breakdown.data.datasets[0].values[i];
            const color = data.charts.margin_breakdown.colors[i % data.charts.margin_breakdown.colors.length];
            const share = ((val / total_val) * 100).toFixed(1) + "%";
            return `
                <div style="display: inline-block; width: 45%; margin: 5px 2%; vertical-align: top; text-align: left;">
                    <span style="display: inline-block; width: 10px; height: 10px; border-radius: 2px; background: ${color}; margin-right: 8px;"></span>
                    <div style="display: inline-block; vertical-align: top;">
                        <div style="font-size: 10px; font-weight: 600; color: #475569;">${label}</div>
                        <div style="font-size: 9px; color: #94a3b8;">${format_currency(val)} (${share})</div>
                    </div>
                </div>
            `;
        }).join('');

		let html = `
            <html>
            <head>
                <style>
                    body { font-family: 'Helvetica', sans-serif; padding: 20px; color: #1e293b; }
                    .header { text-align: center; border-bottom: 3px solid #10b981; padding-bottom: 20px; margin-bottom: 30px; }
                    .kpi-wrapper { display: flex; justify-content: space-between; margin-bottom: 30px; }
                    .kpi-card { border: 1px solid #e2e8f0; padding: 15px; border-radius: 10px; flex: 1; margin: 0 10px; text-align: center; background: #f8fafc; }
                    .kpi-label { font-size: 10px; color: #64748b; font-weight: 700; text-transform: uppercase; margin-bottom: 5px; }
                    .kpi-value { font-size: 18px; font-weight: 800; }
                    .chart-container { text-align: center; margin-bottom: 40px; border: 1px solid #e2e8f0; border-radius: 12px; padding: 24px; background: #fff; page-break-inside: avoid; }
                    .chart-img { max-width: 400px; margin-bottom: 20px; }
                    .pdf-legend-box { background: #fafafa; border-radius: 8px; padding: 15px; border-top: 1px solid #f1f5f9; margin-top: 20px; text-align: left; }
                    
                    table { width: 100%; border-collapse: collapse; font-size: 10px; margin-top: 20px; page-break-inside: auto; }
                    tr { page-break-inside: avoid; page-break-after: auto; }
                    thead { display: table-header-group; }
                    tfoot { display: table-row-group; }
                    th { background: #f1f5f9; padding: 8px; text-align: left; border-bottom: 2px solid #10b981; color: #64748b; text-transform: uppercase; }
                    td { padding: 8px; border-bottom: 1px solid #eee; }
                    .text-right { text-align: right; }
                    .text-center { text-align: center; }
                </style>
            </head>
            <body>
                <div class="header">
                    <h1 style="margin:0;">Margin Build Report (Million INR)</h1>
                    <p style="font-size:10px; color:#999;">Generated: ${report_date}</p>
                </div>
                <div class="kpi-wrapper">
                    ${data.summary.map(m => `
                        <div class="kpi-card">
                            <div class="kpi-label">${m.label}</div>
                            <div class="kpi-value">${m.fieldtype === 'Currency' ? format_currency(m.value) : m.value}</div>
                        </div>
                    `).join('')}
                </div>
                
                <div class="chart-container">
                    <h3 style="color:#334155; text-transform:uppercase; font-size:14px; margin-top:0;">${data.charts.margin_breakdown.title}</h3>
                    <img src="${chart_png}" class="chart-img">
                    <div class="pdf-legend-box">
                        ${legend_box_html}
                    </div>
                </div>

                <h3 style="color:#334155; text-transform:uppercase; font-size:14px;">Detailed Margin List</h3>
                <table>
                    <thead>
                        <tr>
                            <th>Invoice ID</th>
                            <th>Date</th>
                            <th>Customer</th>
                            <th class="text-center">Type</th>
                            <th class="text-right">Revenue (M)</th>
                            <th class="text-right">COGS (M)</th>
                            <th class="text-right">Margin (M)</th>
                        </tr>
                    </thead>
                    <tbody>
                        ${data.results.map(row => `
                            <tr>
                                <td>${row.name}</td>
                                <td>${frappe.datetime.str_to_user(row.posting_date)}</td>
                                <td>${row.customer_name || row.customer}</td>
                                <td class="text-center">${row.type}</td>
                                <td class="text-right">${format_currency(row.revenue)}</td>
                                <td class="text-right">${format_currency(row.cogs)}</td>
                                <td class="text-right" style="font-weight: bold; color: ${row.margin < 0 ? '#ef4444' : '#10b981'}">${format_currency(row.margin)}</td>
                            </tr>
                        `).join('')}
                    </tbody>
                </table>
            </body>
            </html>
        `;

		const method_url = "/api/method/renu_customization.renu_customization.page.margin_build.margin_build.export_to_pdf";
		const $form = $(`<form action="${method_url}" method="POST" target="_blank" style="display:none;">
            <input type="hidden" name="html" value="">
            <input type="hidden" name="csrf_token" value="${frappe.csrf_token}">
        </form>`).appendTo("body");

		$form.find('input[name="html"]').val(html);
		$form.submit();
		$form.remove();
	};

    page.add_menu_item(__("Export to PDF"), () => export_pdf());
    page.add_menu_item(__("Export to Excel"), () => export_to_excel());

	let filter_area = $('<div class="dashboard-filter-area border-bottom"></div>').prependTo(page.main);
	page.container = $('<div class="dashboard-content"></div>').appendTo(page.main);

	const filter_fields = [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_months(frappe.datetime.get_today(), -1)
		},
		{ fieldtype: "Column Break" },
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today()
		},
		{ fieldtype: "Column Break" },
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{ fieldtype: "Column Break" },
		{
			fieldname: "type",
			label: __("Type"),
			fieldtype: "Select",
			options: ["", "Domestic", "Export"]
		}
	];

	page.filter_group = new frappe.ui.FieldGroup({
		parent: filter_area,
		fields: filter_fields,
	});
	page.filter_group.make();

	filter_area.css({
		"padding": "10px 20px",
		"background-color": "#fff",
		"border-bottom": "1px solid #f1f5f9"
	});

	Object.keys(page.filter_group.fields_dict).forEach(key => {
		let field = page.filter_group.fields_dict[key];
		field.df.on_change = () => page.refresh();
		if (field.$input) {
			field.$input.on("change input blur", () => { setTimeout(() => page.refresh(), 50); });
		}
	});

	page.refresh = function() {
		let filters = page.filter_group.get_values();
		
		frappe.call({
			method: "renu_customization.renu_customization.page.margin_build.margin_build.get_dashboard_data",
			args: { filters: filters },
			callback: function(r) {
				if (r.message) {
					page.dashboard_data = r.message;
					render_dashboard(r.message);
				}
			}
		});
	};

	function render_dashboard(data) {
		page.container.empty();

		if (!$('#margin-dashboard-style').length) {
			$(`<style id="margin-dashboard-style">
                @import url('https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700;800&display=swap');
                
                .dashboard-content { 
                    padding: 20px; 
                    background: transparent !important; 
                    min-height: 100vh; 
                    font-family: 'Inter', sans-serif; 
                    color: #1e293b; 
                    width: 100% !important; 
                }

				.summary-wrapper { 
                    display: grid !important; 
                    grid-template-columns: repeat(4, 1fr) !important; 
                    gap: 16px; 
                    margin-bottom: 24px; 
                    width: 100% !important; 
                }
				.summary-card { 
					background: #ffffff; 
                    border: 1px solid #e2e8f0; 
                    border-radius: 12px; 
                    padding: 16px; 
                    box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05); 
                    border-left: 4px solid #cbd5e1; 
                    transition: transform 0.2s;
				}
                .summary-card:hover { transform: translateY(-2px); }
				.summary-card.border-red { border-left-color: #ef4444; }
				.summary-card.border-orange { border-left-color: #f59e0b; }
				.summary-card.border-blue { border-left-color: #3b82f6; }
				.summary-card.border-green { border-left-color: #10b981; }
				.summary-card.border-purple { border-left-color: #8b5cf6; }
				
                .summary-card .label { 
                    font-size: 11px; 
                    color: #64748b; 
                    font-weight: 700; 
                    text-transform: uppercase; 
                    letter-spacing: 0.05em; 
                    margin-bottom: 8px;
                }
				.summary-card .value { 
                    font-size: 20px; 
                    font-weight: 800; 
                    color: #0f172a; 
                }

				.chart-card { 
                    background: #fff; 
                    border-radius: 12px; 
                    padding: 24px; 
                    box-shadow: 0 1px 3px rgba(0,0,0,0.1); 
                    border: 1px solid #e2e8f0; 
                    margin-bottom: 24px;
                }
                .chart-card .title { 
                    font-size: 13px; 
                    font-weight: 700; 
                    color: #1e293b; 
                    margin-bottom: 20px; 
                    text-transform: uppercase; 
                    letter-spacing: 0.05em; 
                }
                
                .custom-legend { 
                    display: grid !important; 
                    grid-template-columns: repeat(auto-fill, minmax(220px, 1fr)) !important; 
                    gap: 12px 24px !important; 
                    margin-top: 20px; 
                    padding: 16px !important; 
                    background: #f8fafc !important; 
                    border-radius: 8px !important; 
                    border: 1px solid #eef2f6 !important; 
                }
                .legend-item { 
                    display: flex !important; 
                    align-items: center !important; 
                    gap: 10px !important; 
                }
                .legend-item .dot { width: 10px !important; height: 10px !important; border-radius: 2px !important; flex-shrink: 0; }
                .legend-item .info { display: flex !important; align-items: center !important; gap: 8px !important; width: 100%; }
                .legend-item .label { font-size: 11px !important; font-weight: 600 !important; color: #475569 !important; white-space: nowrap; }
                .legend-item .val-pct { font-size: 10px !important; color: #94a3b8 !important; }
                .legend-item .val-amount { font-size: 11px !important; font-weight: 700 !important; color: #1e293b !important; margin-left: auto; }
				
				.table-card { 
                    background: #fff; 
                    border-radius: 12px; 
                    box-shadow: 0 1px 3px rgba(0,0,0,0.1); 
                    margin-bottom: 24px; 
                    overflow: hidden; 
                    border: 1px solid #e2e8f0; 
                    width: 100%; 
                }
				.table-card .header { 
                    padding: 15px 24px; 
                    background: #fff; 
                    border-bottom: 1px solid #f1f5f9; 
                    font-weight: 700; 
                    color: #0f172a; 
                    display: flex; 
                    justify-content: space-between; 
                    align-items: center; 
                }
                .table-container { 
                    overflow: auto; 
                    width: 100%; 
                    max-height: 600px; 
                }
				.dashboard-table { width: 100%; border-collapse: separate; border-spacing: 0; }
				.dashboard-table th { 
                    background: #f8fafc; 
                    padding: 12px 16px; 
                    text-align: left; 
                    font-size: 11px; 
                    font-weight: 700; 
                    color: #64748b; 
                    position: sticky; 
                    top: 0; 
                    z-index: 10; 
                    border-bottom: 1px solid #e2e8f0; 
                    text-transform: uppercase; 
                }
				.dashboard-table td { 
                    padding: 12px 16px; 
                    border-bottom: 1px solid #f1f5f9; 
                    font-size: 13px; 
                    color: #334155; 
                    background: #fff; 
                }
				.dashboard-table tr:hover td { background: #f8fafc; }
				.text-right { text-align: right; }
                .text-center { text-align: center; }
				
                .indicator-pill {
                    padding: 4px 10px;
                    border-radius: 12px;
                    font-size: 10px;
                    font-weight: 700;
                    text-transform: uppercase;
                }
                .indicator-pill.Domestic { background: #e0f2fe; color: #0369a1; }
                .indicator-pill.Export { background: #fef3c7; color: #92400e; }
                .indicator-pill.Other { background: #f1f5f9; color: #475569; }

                .export-btn { 
                    font-size: 12px; cursor: pointer; color: #475569; font-weight: 600; 
                    padding: 6px 14px; border-radius: 6px; transition: all 0.2s;
                    display: inline-flex; align-items: center; gap: 6px;
                    background: #fff; border: 1px solid #e2e8f0;
                    white-space: nowrap;
                }
                .export-btn:hover { color: #2563eb; background: #eff6ff; border-color: #bfdbfe; }
			</style>`).appendTo('head');
		}

		let summary_row = $('<div class="summary-wrapper"></div>').appendTo(page.container);
		data.summary.forEach(s => {
			$(`
				<div class="summary-card border-${s.indicator.toLowerCase()}">
					<div class="label">${s.label}</div>
					<div class="value">${s.fieldtype === 'Currency' ? format_currency(s.value) : s.value}</div>
				</div>
			`).appendTo(summary_row);
		});

		let chart_card = $(`
            <div class="chart-card">
                <div class="title">${data.charts.margin_breakdown.title}</div>
                <div id="margin-chart" style="height: 350px;"></div>
                <div id="margin-legend" class="custom-legend"></div>
            </div>
        `).appendTo(page.container);
        
		setTimeout(() => {
            let chart = new frappe.Chart("#margin-chart", {
                data: data.charts.margin_breakdown.data,
                type: 'donut',
                height: 350,
                colors: data.charts.margin_breakdown.colors,
                legend: 0,
            });

            // Render Custom Legend
            let legend_container = chart_card.find("#margin-legend");
            let total_val = data.charts.margin_breakdown.data.datasets[0].values.reduce((a, b) => a + b, 0);

            data.charts.margin_breakdown.data.labels.forEach((label, idx) => {
                let val = data.charts.margin_breakdown.data.datasets[0].values[idx];
                let color = data.charts.margin_breakdown.colors[idx % data.charts.margin_breakdown.colors.length];
                let share = total_val > 0 ? ((val / total_val) * 100).toFixed(1) + "%" : "0%";
                
                legend_container.append(`
                    <div class="legend-item">
                        <span class="dot" style="background: ${color}"></span>
                        <div class="info">
                            <span class="label">${label}</span>
                            <span class="val-pct">(${share})</span>
                            <span class="val-amount">${format_currency(val)}</span>
                        </div>
                    </div>
                `);
            });
        }, 100);

		let table_card = $(`
            <div class="table-card">
                <div class="header">
                    <span>${__("Detailed Margin List")}</span>
                    <div class="export-btn" id="export_excel_table">
                        <i class="fa fa-file-excel-o"></i> Export
                    </div>
                </div>
                <div class="table-container">
                    <table class="dashboard-table">
                        <thead>
                            <tr>
                                <th>${__("Invoice ID")}</th>
                                <th>${__("Date")}</th>
                                <th>${__("Customer")}</th>
                                <th class="text-center">${__("Type")}</th>
                                <th class="text-right">${__("Revenue (M)")}</th>
                                <th class="text-right">${__("COGS (M)")}</th>
                                <th class="text-right">${__("Margin (M)")}</th>
                            </tr>
                        </thead>
                        <tbody></tbody>
                    </table>
                </div>
            </div>
        `).appendTo(page.container);

		let tbody = table_card.find('tbody');
		data.results.forEach(row => {
			$(`
				<tr>
					<td><a href="/app/sales-invoice/${row.name}" style="font-weight: 600; color: #4338ca;">${row.name}</a></td>
					<td style="white-space: nowrap;">${frappe.datetime.str_to_user(row.posting_date)}</td>
					<td style="font-weight: 500;">${row.customer_name || row.customer}</td>
                    <td class="text-center">
                        <span class="indicator-pill ${row.type}">${__(row.type)}</span>
                    </td>
					<td class="text-right">${format_currency(row.revenue)}</td>
					<td class="text-right">${format_currency(row.cogs)}</td>
					<td class="text-right" style="font-weight: 700; color: ${row.margin < 0 ? '#ef4444' : '#10b981'};">${format_currency(row.margin)}</td>
				</tr>
			`).appendTo(tbody);
		});

        table_card.find("#export_excel_table").click(() => export_to_excel());
	}

	function format_currency(v) {
		if (!v && v !== 0) return "₹ 0.00 M";
		
		let value = flt(v) / 1000000;
		
		return (
			"₹ " +
			value.toLocaleString("en-US", {
				minimumFractionDigits: 2,
				maximumFractionDigits: 2,
			}) + " M"
		);
	}

	page.refresh();
};
